File: src/main/kotlin/cc/mcyx/paimon/common/ui/Sign.java
package cc.mcyx.paimon.common.ui;

import cc.mcyx.paimon.common.minecraft.craftbukkit.CraftBukkitPacket;
import cc.mcyx.paimon.common.minecraft.network.PaimonPlayer;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;

public class Sign {
    static Class<?> BlockPosition = CraftBukkitPacket.Companion.asNMSPacketClass("BlockPosition");
    static Class<?> Blocks = CraftBukkitPacket.Companion.asNMSPacketClass("Blocks");
    static Class<?> PacketPlayOutBlockChange = CraftBukkitPacket.Companion.asNMSPacketClass("PacketPlayOutBlockChange");
    static Class<?> PacketPlayOutTileEntityData = CraftBukkitPacket.Companion.asNMSPacketClass("PacketPlayOutTileEntityData");
    static Class<?> NBTTagCompound = CraftBukkitPacket.Companion.asNMSPacketClass("NBTTagCompound");
    static Class<?> PacketPlayOutOpenSignEditor = CraftBukkitPacket.Companion.asNMSPacketClass("PacketPlayOutOpenSignEditor");

    //117
    static Class<?> IBlockData = CraftBukkitPacket.Companion.asNMSPacketClass("IBlockData");
    static Class<?> TileEntityTypes = CraftBukkitPacket.Companion.asNMSPacketClass("TileEntityTypes");
    static Class<?> MojangsonParser = CraftBukkitPacket.Companion.asNMSPacketClass("MojangsonParser");

    //112
    static Class<?> CraftWorld = CraftBukkitPacket.Companion.asBukkitClass("CraftWorld");
    static Class<?> Entity = CraftBukkitPacket.Companion.asNMSPacketClass("Entity");
    static Class<?> World = CraftBukkitPacket.Companion.asNMSPacketClass("World");

    public static void openSign(Player player, String[] texts) {
        try {
            PaimonPlayer paimonPlayer = new PaimonPlayer(player);
            if (CraftBukkitPacket.Companion.getServerId() >= 1170) {
                //创建坐标数据包
                Block block = player.getLocation().getBlock();
                Object bp = BlockPosition.getConstructor(int.class, int.class, int.class).newInstance(block.getX(), block.getY(), block.getZ());

                //创建牌子数据包
                Object STANDING_SIGN = Blocks.getField("cE").get(Blocks);
                Object bc = PacketPlayOutBlockChange.getConstructor(BlockPosition, IBlockData).newInstance(bp,
                        STANDING_SIGN.getClass().getMethod("n").invoke(STANDING_SIGN));
                paimonPlayer.sendPacket(bc);

                //改变牌子文本数据包
                Object tileEntityTypes = TileEntityTypes.getField("h").get(TileEntityTypes);
                Constructor<?> constructor = PacketPlayOutTileEntityData.getDeclaredConstructor(BlockPosition, TileEntityTypes, NBTTagCompound);
                constructor.setAccessible(true);

                String s = String.format("{front_text:{messages:['{\"text\":\"%s\"}','{\"text\":\"%s\"}','{\"text\":\"%s\"}','{\"text\":\"%s\"}']}}", texts[0], texts[1], texts[2], texts[3]);
                Object packetSignTextData = MojangsonParser.getMethod("a", String.class).invoke(MojangsonParser, s);

                Object packetPlayOutTileEntityData = constructor.newInstance(bp, tileEntityTypes, packetSignTextData);
                paimonPlayer.sendPacket(packetPlayOutTileEntityData);

                //打开牌子数据包
                paimonPlayer.sendPacket(PacketPlayOutOpenSignEditor.getConstructor(BlockPosition, boolean.class).newInstance(bp, true));

                //移除牌子数据包
                Object AIR = Blocks.getField("a").get(Blocks);
                bc = PacketPlayOutBlockChange.getConstructor(BlockPosition, IBlockData).newInstance(bp, AIR.getClass().getMethod("n").invoke(AIR));
                paimonPlayer.sendPacket(bc);
            }  else if (CraftBukkitPacket.Companion.getServerId() == 1165) {
                //创建坐标数据包
                Block block = player.getLocation().getBlock();
                Object bp = BlockPosition.getConstructor(int.class, int.class, int.class).newInstance(block.getX(), block.getY(), block.getZ());

                //创建牌子数据包
                Object STANDING_SIGN = Blocks.getField("OAK_SIGN").get(Blocks);
                Object bc = PacketPlayOutBlockChange.getConstructor(BlockPosition, IBlockData).newInstance(bp,
                        STANDING_SIGN.getClass().getMethod("getBlockData").invoke(STANDING_SIGN));
                paimonPlayer.sendPacket(bc);

                Object packetSignTextData = NBTTagCompound.getConstructor().newInstance();
                Method method = packetSignTextData.getClass().getMethod("setString", String.class, String.class);
                method.invoke(packetSignTextData, "id", "minecraft:sign");
                method.invoke(packetSignTextData, "Text1", String.format("{\"text\":\"%s\"}", texts[0]));
                method.invoke(packetSignTextData, "Text2", String.format("{\"text\":\"%s\"}", texts[1]));
                method.invoke(packetSignTextData, "Text3", String.format("{\"text\":\"%s\"}", texts[2]));
                method.invoke(packetSignTextData, "Text4", String.format("{\"text\":\"%s\"}", texts[3]));
                Method method1 = packetSignTextData.getClass().getMethod("setInt", String.class, int.class);
                method1.invoke(packetSignTextData, "x", block.getX());
                method1.invoke(packetSignTextData, "y", block.getY());
                method1.invoke(packetSignTextData, "z", block.getZ());

                //改变牌子文本数据包
                Object packetPlayOutTileEntityData = PacketPlayOutTileEntityData.getConstructor(BlockPosition, int.class, NBTTagCompound).newInstance(bp, 9, packetSignTextData);
                paimonPlayer.sendPacket(packetPlayOutTileEntityData);

                //打开牌子数据包
                paimonPlayer.sendPacket(PacketPlayOutOpenSignEditor.getConstructor(BlockPosition).newInstance(bp));
            } else {
                //创建坐标数据包
                Object world = CraftWorld.cast(player.getWorld());
                System.out.println(Arrays.toString(BlockPosition.getConstructors()));

                Object bp = BlockPosition.getConstructor(Entity).newInstance(paimonPlayer.getEntityPlayer());

                //创建牌子数据包
                Object bc = PacketPlayOutBlockChange.getConstructor(World, BlockPosition).newInstance(
                        world.getClass().getMethod("getHandle").invoke(world), bp);
                Object STANDING_SIGN = Blocks.getField("STANDING_SIGN").get(Blocks);
                bc.getClass().getField("block").set(bc,
                        STANDING_SIGN.getClass().getMethod("getBlockData").invoke(STANDING_SIGN));
                paimonPlayer.sendPacket(bc);

                Object packetSignTextData = NBTTagCompound.getConstructor().newInstance();
                Method method = packetSignTextData.getClass().getMethod("setString", String.class, String.class);
                method.invoke(packetSignTextData, "id", "minecraft:sign");
                method.invoke(packetSignTextData, "Text1", String.format("{\"text\":\"%s\"}", texts[0]));
                method.invoke(packetSignTextData, "Text2", String.format("{\"text\":\"%s\"}", texts[1]));
                method.invoke(packetSignTextData, "Text3", String.format("{\"text\":\"%s\"}", texts[2]));
                method.invoke(packetSignTextData, "Text4", String.format("{\"text\":\"%s\"}", texts[3]));

                //改变牌子文本数据包
                Object packetPlayOutTileEntityData = PacketPlayOutTileEntityData.getConstructor(BlockPosition, int.class, NBTTagCompound).newInstance(bp, 9, packetSignTextData);
                paimonPlayer.sendPacket(packetPlayOutTileEntityData);

                //打开牌子数据包
                paimonPlayer.sendPacket(PacketPlayOutOpenSignEditor.getConstructor(BlockPosition).newInstance(bp));

                //移除牌子数据包
                Object AIR = Blocks.getField("AIR").get(Blocks);
                bc.getClass().getField("block").set(bc, AIR.getClass().getMethod("getBlockData").invoke(AIR));
                paimonPlayer.sendPacket(bc);
            }
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException |
                 NoSuchFieldException e) {
            throw new RuntimeException(e);
        }
    }
}

File: src/main/kotlin/cc/mcyx/paimon/common/ui/Sign.kt
package cc.mcyx.paimon.common.ui

import cc.mcyx.paimon.common.minecraft.craftbukkit.CraftBukkitPacket
import cc.mcyx.paimon.common.minecraft.craftbukkit.CraftBukkitPacket.Companion.serverId
import cc.mcyx.paimon.common.minecraft.network.PaimonPlayer
import org.bukkit.entity.Player

class Sign(private val player: Player) {
    private companion object {
        /*
        * 共有类(1.8-最新版本)
        * */
        val BlockPosition = CraftBukkitPacket.asNMSPacketClass("BlockPosition")
        val PacketPlayOutBlockChange = CraftBukkitPacket.asNMSPacketClass("PacketPlayOutBlockChange")
        val Blocks = CraftBukkitPacket.asNMSPacketClass("Blocks")
        val Block = CraftBukkitPacket.asNMSPacketClass("Block")
        val PacketPlayOutTileEntityData = CraftBukkitPacket.asNMSPacketClass("PacketPlayOutTileEntityData")
        val NBTTagCompound = CraftBukkitPacket.asNMSPacketClass("NBTTagCompound")
        val PacketPlayOutOpenSignEditor = CraftBukkitPacket.asNMSPacketClass("PacketPlayOutOpenSignEditor")
        val PacketPlayInUpdateSign = CraftBukkitPacket.asNMSPacketClass("PacketPlayInUpdateSign")
        val MojangsonParser = CraftBukkitPacket.asNMSPacketClass("MojangsonParser")

        /*
        * 特殊类
        * */
        val ChatComponentText = CraftBukkitPacket.asNMSPacketClass("ChatComponentText")  //1.10以下
        val IChatBaseComponent = CraftBukkitPacket.asNMSPacketClass("IChatBaseComponent")  //1.10以下
        val World = CraftBukkitPacket.asNMSPacketClass("World")  //1.10以下
        val PacketPlayOutUpdateSign = CraftBukkitPacket.asNMSPacketClass("PacketPlayOutUpdateSign")  //1.10以下
        val IBlockData = CraftBukkitPacket.asNMSPacketClass("IBlockData")  //1.17以上
        val TileEntityTypes = CraftBukkitPacket.asNMSPacketClass("TileEntityTypes")  //1.17以上

        /*
        * 静态字段
        * */
        val sign = if (serverId >= 1170) "cE" else if (serverId >= 1165) "OAK_SIGN" else "STANDING_SIGN"
        val signIBlockData = CraftBukkitPacket.getStaticObject(Block, Blocks.getField(sign).get(null), "IBlockData")
    }

    /*
    * 显示牌子编辑界面并显示文字
    * */
    fun open(line: Array<String> = arrayOf()): Sign {
        line.forEachIndexed { index, s ->
            line[index] = s.replace("&","§")
        }
        val paimonPlayer = PaimonPlayer(player)
        val block = player.location.block
        val blockPosition = BlockPosition.getConstructor(
            Int::class.java, Int::class.java, Int::class.java
        ).newInstance(block.x, block.y, block.z)

        //创建改变方块数据包(改变成牌子)
        val packetPlayOutBlockChange =
            if (serverId >= 1170)
                PacketPlayOutBlockChange.getDeclaredConstructor(BlockPosition, IBlockData)
                    .newInstance(blockPosition, signIBlockData)
            else PacketPlayOutBlockChange.newInstance().also {
                it.javaClass.getDeclaredField(CraftBukkitPacket.getObjectName(it, "BlockPosition"))
                    .apply { isAccessible = true }.set(it, blockPosition)
                it.javaClass.getDeclaredField(CraftBukkitPacket.getObjectName(it, "IBlockData"))
                    .set(it, signIBlockData)
            }
        paimonPlayer.sendPacket(packetPlayOutBlockChange)

        //创建编辑牌子数据包
        if (serverId >= 1100) {
            val nBTTagCompound =
                MojangsonParser.getMethod(if (serverId >= 1180) "a" else "parse", String::class.java).invoke(
                    null,
                    if (serverId >= 1180) """
                    {front_text:{messages:['
                    {"text":"${if (line.isNotEmpty()) line[0] else ""}"}',
                    '{"text":"${if (line.size >= 2) line[1] else ""}"}',
                    '{"text":"${if (line.size >= 3) line[2] else ""}"}',
                    '{"text":"${if (line.size >= 4) line[3] else ""}"}']}}
                """.trimIndent()
                    else """
                    {Text1:"{\"text\":\"${if (line.isNotEmpty()) line[0] else ""}\"}",
                    Text2:"{\"text\":\"${if (line.size >= 2) line[1] else ""}\"}",
                    Text3:"{\"text\":\"${if (line.size >= 3) line[2] else ""}\"}",
                    Text4:"{\"text\":\"${if (line.size >= 4) line[3] else ""}\"}"}
                """.trimIndent()
                )
            val packetPlayOutTileEntityData =
                if (serverId >= 1180)
                    PacketPlayOutTileEntityData.getDeclaredConstructor(BlockPosition, TileEntityTypes, NBTTagCompound)
                        .apply { isAccessible = true }
                        .newInstance(blockPosition, TileEntityTypes.getDeclaredField("h").get(null), nBTTagCompound)
                else PacketPlayOutTileEntityData.getConstructor(BlockPosition, Int::class.java, NBTTagCompound)
                    .newInstance(blockPosition, 9, nBTTagCompound)
            paimonPlayer.sendPacket(packetPlayOutTileEntityData)
        } else {
            val chatComponents = java.lang.reflect.Array.newInstance(IChatBaseComponent, 4)
            line.forEachIndexed { index, s ->
                val chatSerializer: Any = IChatBaseComponent.declaredClasses[0].newInstance()
                val o = chatSerializer.javaClass.getMethod("a", String::class.java)
                    .invoke(chatSerializer, "{\"text\":\"${if (line.size >= index + 1) s else ""}\"}")
                java.lang.reflect.Array.set(chatComponents, index, o)
            }
            val packetPlayOutUpdateSign: Any = PacketPlayOutUpdateSign.getConstructor(
                World, BlockPosition,
                chatComponents.javaClass
            ).newInstance(paimonPlayer.getNMSWorld(), blockPosition, chatComponents)
            paimonPlayer.sendPacket(packetPlayOutUpdateSign)
        }

        //创建打开牌子数据包
        val packetPlayOutOpenSignEditor =
            if (serverId >= 1200)
                PacketPlayOutOpenSignEditor.getConstructor(BlockPosition, Boolean::class.java)
                    .newInstance(blockPosition, true)
            else PacketPlayOutOpenSignEditor.getConstructor(BlockPosition).newInstance(blockPosition)
        paimonPlayer.sendPacket(packetPlayOutOpenSignEditor)

        var first = true
        paimonPlayer.packetListener {
            if (first && it.packet.javaClass.simpleName == "PacketPlayInUpdateSign") {
                first = false
                val packetPlayInUpdateSign = PacketPlayInUpdateSign.cast(it.packet)
                val lineValue = (CraftBukkitPacket.getStaticObject(
                    PacketPlayInUpdateSign,
                    packetPlayInUpdateSign,
                    if (serverId >= 1100) "String;" else "IChatBaseComponent;"
                ) as Array<*>).toMutableList()
                if (serverId < 1100)
                    repeat(lineValue.size) { i ->
                        lineValue[i] = ChatComponentText.cast(lineValue[i]).javaClass.getDeclaredField("b")
                            .apply { isAccessible = true }.get(lineValue[i])
                    }
                this.line?.invoke(player, lineValue)
                paimonPlayer.player.sendBlockChange(
                    player.location,
                    player.location.block.type,
                    player.location.block.data
                )
                it.isCancel = true
            }
        }

        return this
    }

    private var line: ((Player, List<*>) -> Unit)? = null

    fun getLine(line: ((Player, List<*>) -> Unit)?) {
        this.line = line
    }

}